src/dune_rpc_server/dune_rpc_server.ml

```diff
@@ -481,29 +481,48 @@ module H = struct
           | Cancelled
       end
 
-      let on_cancel map _session poller =
+      let fill_if_possible ivar value =
+        let* current = Fiber.Ivar.peek ivar in
+        match current with
+        | Some _ -> Fiber.return ()
+        | None -> Fiber.Ivar.fill ivar value
+
+      let on_cancel map pending _session poller =
         let new_map =
           Map.update !map poller ~f:(function
             | None -> assert false
             | Some Status.Cancelled as s -> s
             | Some (Active _) -> Some Cancelled)
         in
         map := new_map;
-        Fiber.return ()
+        match !pending with
+        | None -> Fiber.return ()
+        | Some ready -> fill_if_possible ready `Cancelled
 
-      let make_on_poll map svar ~equal ~diff _session poller =
+      let make_on_poll map svar pending ~equal ~diff (session : _ Session.t)
+          poller =
         let send last =
-          let* () =
-            match last with
-            | None -> Fiber.return ()
-            | Some last ->
-              let until x = not (equal x last) in
-              Fiber.Svar.wait svar ~until
+          let ready = Fiber.Ivar.create () in
+          pending := Some ready;
+          let wait_for_svar () =
+            let* () =
+              match last with
+              | None -> Fiber.return ()
+              | Some last ->
+                let until x = not (equal x last) in
+                Fiber.Svar.wait svar ~until
+            in
+            fill_if_possible ready `Data_available
           in
-          let now = Fiber.Svar.read svar in
-          map := Map.set !map poller (Status.Active now);
-          let to_send = diff ~last ~now in
-          Fiber.return (Some to_send)
+          let* () = Fiber.Pool.task session.base.pool ~f:wait_for_svar in
+          let* result_ready = Fiber.Ivar.read ready in
+          match result_ready with
+          | `Cancelled -> Fiber.return None
```

Comment:
Shouldn't we remove the poller on cancellation as well?

Reply:
Good catch. I think it doesn't actually matter, because we only wait on `!result_ready` after a new poll is initiated (which means that we'd have re-filled it with a fresh ivar), but it's a good idea to remove it anyway.

Reply:
Yes, I think so too. Otherwise it leaks a little memory as well.